### PR TITLE
Add typescript providers

### PR DESCRIPTION
**Pull request by ZacLN**
Adds TS providers that talk to a single Julia process for:
- Hovers
- Documentation
- Completions
- Linting
- Function definitions

Only tested on Linux w/ vscode 1.6.1 and recent Julia version.

I've done it with a fairly even TS/Julia code mix as implementing the pure julia language-server  meant that one couldn't provide these features for unsaved files (i.e. the client-server interface will only transfer file location details not individual snippets of code)

### CODE REVIEW

**Comment by davidanthoff** [comment]
Oh, one more thing: are you the creator of this code? I saw some emails that suggested that someone copied the code from somewhere, but it was not clear to me whether you (@ZacLN) are the person who did the copying or the original author of the code.

**Comment by ZacLN** [comment]
Ok, alot to reply to. I'm inexperienced with contributing so I thought I'd dump out what I had first so that you could decide how to incorporate it/h

**Comment by davidanthoff** [comment]
Yeah, no worries, this is actually useful because it gives me a really good overview where this is going! Plus it is just fantastic how much you created here. Happy to work together to get this merged eventually.

**Comment by ZacLN** [comment]
Sorry, pressed send by mistake, to continue:
 
- Re code origins: nova/novatena2015 appeared to have an updated fork which has since disappeared. This had the julia proxy back-end and code completion,hovers and definition goto. This build on the previous package.
- Re language server vs TS/julia mix: My main issue with the pure language-server approach was that working on unsaved files would not work and that working with unsaved changes would require some short term version control system on the Julia side. Though doable this seems to require quite a bit more work for little practical gain (I tried this initially because I thought it a more elegant solution). Client side requests for things like hover/completions just send selection position data.
- In terms of package setup: I don't have access to Mac/Windows systems so would need some OS specific help in how to do this across platforms. As an initial approach would you think a fixed path in the settings contribution, similar to that given for the package location,  appropriate?

Once you give me an idea of how you think it best to progress with all this I'll start again with smaller PRs

Ta, Zac

File: jl/server.jl [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+    "symbols" => symbolrequesthandler,
+    "diagnostics" => diagnosticshandler
+)
+
+function run()
+    while true
+        request = nothing
+        ok = false
+        try
+            request = JSON.parse(readline())
```
Comment (ZacLN) [inline jl/server.jl]: This doesn't seem to matter: things printed to STDOUT not fitting the expected form of a ?Result don't get picked up by the TS side of things.

File: jl/utils.jl [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+    str = code[1:pos]
+    return split(str, '\n')
+end
+
+function isimportcmd(cmd::AbstractString)
+    return startswith(cmd, "using") || startswith(cmd, "import")
+end
+
+function doimport!(cmdcache::Set{AbstractString}, filename::AbstractString)
+    lines = readfirstlines(filename, 20)
```
Comment (ZacLN) [inline jl/utils.jl]: This is code I didn't write/change. I was going to change it to simply search for all instances of using/import/include to ensure all relevant modules get included into the namespace.

File: src/async.ts [hunk 3]
```diff
@@ -1,187 +0,0 @@
-/*---------------------------------------------------------------------------------------------
```
Comment (ZacLN) [inline src/async.ts]: This was used for the previous linter, not sure of it's exact role but it seems easier to keep all the components running on one process?

**Pull request closed without merging**